### PR TITLE
Request PoW target difficulty for miner from the base node

### PR DESCRIPTION
**Pull request by neonknight64**
## Description
- Introduced the difficulty adjustment manager that keeps track of the chain state to calculate the target difficulty of the different PoW algorithms.
- Added the GetTargetDifficulty comms request and TargetDifficulty comms response to the BaseNodeService allowing miners to request the target difficulty from a Base node.
- Added proto conversions for the GetTargetDifficulty request and TargetDifficulty response.
- Added the difficulty adjustment manager to the tari base node and existing tests.
- Fixed 0 difficulty issue in Lwma_diff and change it to make use of the correct target time. Changed it to make the target time configurable.

## Motivation and Context
Enables miners to query a Base node to determine the target difficulty for a specified PoW algorithm.

## How Has This Been Tested?
Added a test for testing the the get target difficulty request and response system and tests for the DiffAdjManager system.

## Types of changes

* [x] Bug fix (non-breaking change which fixes an issue)
* [x] New feature (non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to change)
* [ ] Feature refactor (No new feature or functional changes, but performance or technical debt improvements)
* [x] New Tests
* [ ] Documentation

## Checklist:

* [x] I'm merging against the `development` branch
* [x] I ran `cargo-fmt --all` before pushing
* [ ] My change requires a change to the documentation.
* [ ] I have updated the documentation accordingly.
* [x] I have added tests to cover my changes.

### CODE REVIEW

File: base_layer/core/src/proof_of_work/lwma_diff.rs [hunk 1]
```diff
@@ -66,23 +68,24 @@ impl LinearWeightedMovingAverage {
...
             weighted_times += solve_time * i as u64;
         }
         // k is the sum of weights (1+2+..+n) * target_time
-        let k = n * (n + 1) * TARGET_BLOCK_INTERVAL / 2;
+        let k = n * (n + 1) * self.target_time / 2;
+        dbg!(k);
         let target = ave_difficulty * k as f64 / weighted_times as f64;
+        dbg!(&target);
+        dbg!(&weighted_times);
+        dbg!(&ave_difficulty);
```
Comment (SWvheerden) [inline base_layer/core/src/proof_of_work/lwma_diff.rs]: remove test code

File: base_layer/core/src/proof_of_work/lwma_diff.rs [hunk 2]
```diff
@@ -66,23 +68,24 @@ impl LinearWeightedMovingAverage {
...
             previous_timestamp = this_timestamp;
 
             // Give linearly higher weight to more recent solve times.
             // Note: This will not overflow for practical values of block_window and solve time.
             weighted_times += solve_time * i as u64;
         }
         // k is the sum of weights (1+2+..+n) * target_time
-        let k = n * (n + 1) * TARGET_BLOCK_INTERVAL / 2;
+        let k = n * (n + 1) * self.target_time / 2;
+        dbg!(k);
```
Comment (SWvheerden) [inline base_layer/core/src/proof_of_work/lwma_diff.rs]: remove test code

File: base_layer/transactions/src/consensus.rs [hunk 3]
```diff
@@ -25,11 +25,16 @@ use chrono::{DateTime, Duration, Utc};
 use std::ops::Add;
 
 // This is the our target time in seconds between blocks
-pub const TARGET_BLOCK_INTERVAL: u64 = 60;
+pub const TARGET_BLOCK_INTERVAL: u64 = 120;
 // When doing difficulty adjustments and FTL calculations this is the amount of blocks we look at
-pub const DIFFICULTY_BLOCK_WINDOW: u64 = 150;
+pub const DIFFICULTY_BLOCK_WINDOW: u64 = 120;
```
Comment (SWvheerden) [inline base_layer/transactions/src/consensus.rs]:
```suggestion
pub const DIFFICULTY_BLOCK_WINDOW: u64 = 90;
```
For t = 120, N should be 90
[https://github.com/zawy12/difficulty-algorithms/issues/3](https://github.com/zawy12/difficulty-algorithms/issues/3)

File: base_layer/core/src/proof_of_work/lwma_diff.rs (hunk 1 above)
Comment (neonknight64) [reply base_layer/core/src/proof_of_work/lwma_diff.rs]: Change applied.

File: base_layer/core/src/proof_of_work/lwma_diff.rs (hunk 2 above)
Comment (neonknight64) [reply base_layer/core/src/proof_of_work/lwma_diff.rs]: Change applied.

File: base_layer/transactions/src/consensus.rs (hunk 3 above)
Comment (neonknight64) [reply base_layer/transactions/src/consensus.rs]: Change applied.

File: base_layer/core/src/proof_of_work/diff_adj_manager/error.rs [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+use crate::{chain_storage::ChainStorageError, proof_of_work::error::DifficultyAdjustmentError};
+use derive_error::Error;
+
+#[derive(Debug, Error, Clone, PartialEq)]
+pub enum DiffAdjManagerError {
+    DifficultyAdjustmentError(DifficultyAdjustmentError),
+    ChainStorageError(ChainStorageError),
+    EmptyBlockchain,
+    PoisonedAccess,
```
Comment (CjS77) [inline base_layer/core/src/proof_of_work/diff_adj_manager/error.rs]: I think I've mentioned this before, but I'd rather not lift RWLock errors to the caller -- there's no real action they can take, right; it's usually a sign of a bug? I'd just bundle this in with the general error case with a suitable message.

File: base_layer/core/src/proof_of_work/diff_adj_manager/diff_adj_storage.rs [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+/// system upto date with the blockchain db.
+enum UpdateState {
+    FullSync,
+    SyncToTip,
+    Synced,
+}
+
+/// DiffAdjManager makes use of DiffAdjStorage to provide thread save access to its LinearWeightedMovingAverages for
+/// each PoW algorithm.
+pub struct DiffAdjStorage<T>
```
Comment (CjS77) [inline base_layer/core/src/proof_of_work/diff_adj_manager/diff_adj_storage.rs]: This is a lot of code for a "keep a running list of the last N block data"-- it works, so we'll leave it as is, but did you consider any other strategies?:

* Use VecDequeues and just update from the DB every time? i.e. assume we're always in `SyncToTip` but you pop off the front and push to the back until you have enough data
* Or make it someone else's job to keep the Queues up to date (like a subscriber to a newBlock message)

**Pull request merged**